### ISSUE TITLE
[pango] build failure

### ISSUE THREAD

**Issue by acromarcopolo** [microsoft/vcpkg#18118]
**Host Environment**
- OS: macOS Big Sur 11.3.1
- Compiler: Apple clang version 12.0.5 (clang-1205.0.22.9)
Target: x86_64-apple-darwin20.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

**To Reproduce**
`./vcpkg install gtk`

**Failure logs** 
```
$ ./vcpkg install gtk
Computing installation plan...
The following packages will be built and installed:
  * cairo[core,fontconfig,freetype,gobject]:x64-osx -> 1.17.4#1
  * fontconfig[core]:x64-osx -> 2.13.1#7
  * fribidi[core]:x64-osx -> 1.0.10#2
  * gdk-pixbuf[core]:x64-osx -> 2.42.2#1
  * getopt[core]:x64-osx -> 0
  * graphene[core]:x64-osx -> 1.10.2#1
    gtk[core]:x64-osx -> 4.3.0
  * harfbuzz[core,coretext]:x64-osx -> 2.8.1
  * libepoxy[core]:x64-osx -> 1.5.5
  * libjpeg-turbo[core]:x64-osx -> 2.0.6
  * liblzma[core]:x64-osx -> 5.2.5#2
  * libsass[core]:x64-osx -> 3.6.4
  * lzo[core]:x64-osx -> 2.10#6
  * pango[core]:x64-osx -> 1.48.4
  * pixman[core]:x64-osx -> 0.40.0
  * sassc[core]:x64-osx -> 3.6.1
  * tiff[core]:x64-osx -> 4.1.0#2
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-osx...
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/1c/1c8cb7de0f682c3f9233bd5de49934902c36d902.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/85/851bbe7e50a91bc622f782155b0717591947ffc2.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/3d/3dbda4b3e6c936aa9dcbc596d6d062e5cfd3f4b0.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/d4/d438e51ae829299ce16f84970de16da4989c69d5.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/0e/0eba9d4a292be9181253c9133262a46a0f9bfc43.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/d6/d6e4696dbf72adefd4ffa95b54d6e23e6345d5ef.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/28/28829289752fa75f4a0bdf8bfec723b0f2793ccd.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/88/88b52934ea6fef26ef66e365f7561e73500a9952.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/45/45d5c8234e07856c9df45572c42a231046679adc.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/ca/cad97a2449c1ab664db231fd0fdbbb135f20f540.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/04/04fc8e24e81b0d1bd7001757ce838fadb2718094.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/98/98feeff06cb1c408f9d680dd8ac22799adf05db1.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/3c/3ccc2e229750b72d36a83d98296f7d9b0e187682.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/69/69729320cfe35c8a972fbbf70943bc533114bb81.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/a0/a09b45c00c560e0cdc20bcaf43417dbd18a014fb.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/47/47051fccd497b409f12e5578ac875f4373082d2b.zip
Could not locate cached archive: /Users/marcbauchet/.cache/vcpkg/archives/bd/bd77bd7c4ce0e82535250d1e1e37c6f5e3fffa9a.zip
Starting package 1/17: fontconfig:x64-osx
Building package fontconfig[core]:x64-osx...
-- Using cached /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/downloads/fontconfig-fontconfig-844d8709a1f3ecab45015b24b72dd775c13b2421.tar.gz
-- Cleaning sources at /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/buildtrees/fontconfig/src/75c13b2421-12bdc6f9c0.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/downloads/fontconfig-fontconfig-844d8709a1f3ecab45015b24b72dd775c13b2421.tar.gz
-- Applying patch remove_tests.patch
-- Applying patch build.patch
-- Applying patch build2.patch
-- Applying patch disable-install-data.patch
-- Applying patch macos_arm_fccache.patch
-- Using source at /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/buildtrees/fontconfig/src/75c13b2421-12bdc6f9c0.clean
-- Configuring x64-osx-dbg
-- Configuring x64-osx-rel
-- Generating configure for x64-osx
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:105 (message):
    Command failed: /usr/local/bin/autoreconf -vfi
    Working Directory: /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/buildtrees/fontconfig/src/75c13b2421-12bdc6f9c0.clean/
    Error code: 2
    See logs for more information:
      /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/buildtrees/fontconfig/autoconf-x64-osx-out.log
      /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/buildtrees/fontconfig/autoconf-x64-osx-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:600 (vcpkg_execute_required_process)
  ports/fontconfig/portfile.cmake:26 (vcpkg_configure_make)
  scripts/ports.cmake:141 (include)


Error: Building package fontconfig:x64-osx failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `./vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: fontconfig:x64-osx
  Vcpkg version: 2021-05-05-unknownhash

```
[stdout-x64-osx.log](https://github.com/microsoft/vcpkg/files/6544481/stdout-x64-osx.log)
[autoconf-x64-osx-err.log](https://github.com/microsoft/vcpkg/files/6544482/autoconf-x64-osx-err.log)

**Additional context**
This is a follow up from issue [gtkmm] build failure (#18084) filed May 24, 2021
in between i used brew to install autoconf and gettext
Now both` ./vcpkg install gtkmm ` and `gtk` give the same error as above

**Comment by NancyLi1013** [comment]
Hi @acromarcopolo 
Thanks for posting this issue.
```
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: running: autopoint --force
autoreconf: running: aclocal --force -I m4
Can't exec "aclocal": No such file or directory at /usr/local/Cellar/autoconf/2.71/share/autoconf/Autom4te/FileUtils.pm line 274.
autoreconf: error: aclocal failed with exit status: 2
```

Can you try to install automake via `sudo apt-get install automake` or `brew install automake` and rebuild this?

**Comment by acromarcopolo** [comment]
hi @NancyLi1013,  thanks for your quick feedback.
ok, `brew install automake` done.
`./vcpkg install gtk` (run in vcpkg folder) fails at the same location, also with `autoreconf` but different error code and log, please see attached files:
[stdout-x64-osx.log](https://github.com/microsoft/vcpkg/files/6546587/stdout-x64-osx.log)
[autoconf-x64-osx-err.log](https://github.com/microsoft/vcpkg/files/6546588/autoconf-x64-osx-err.log)
 should i run some kind of `autoupdate` ?  with what parameters, where to run it, etc..?

**Comment by NancyLi1013** [comment]
Can you try to use the suggestion here https://stackoverflow.com/questions/18978252/error-libtool-library-used-but-libtool-is-undefined?

**Comment by acromarcopolo** [comment]
you mean doing an `install libtool`?  ok so i did that, and seems like i am going a bit further:
[stdout-x64-osx.log](https://github.com/microsoft/vcpkg/files/6556092/stdout-x64-osx.log)
[package-x64-osx-dbg-out.log](https://github.com/microsoft/vcpkg/files/6556093/package-x64-osx-dbg-out.log)

so i checked that mysterious ninja, no idea what's it supposed to be/do but i noticed this difference: 
```
% which ninja
/Users/marcbauchet/.new_local/bin/ninja
% diff /Users/marcbauchet/.new_local/bin/ninja /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/downloads/tools/ninja-1.10.1-osx/ninja
Binary files /Users/marcbauchet/.new_local/bin/ninja and /Users/marcbauchet/Dropbox/projects/RandomBits/vcpkg/downloads/tools/ninja-1.10.1-osx/ninja differ
```
any idea?  new issue?

**Comment by NancyLi1013** [comment]
Hi @acromarcopolo 

This is not related with Ninja.
```
../src/480391f31f-4c049f225c.clean/pango/pangofc-fontmap.c:1778:10: error: implicit declaration of function 'FcWeightFromOpenTypeDouble' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  return FcWeightFromOpenTypeDouble (pango_weight);
         ^
../src/480391f31f-4c049f225c.clean/pango/pangofc-fontmap.c:2535:3: error: implicit declaration of function 'FcStrListFirst' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  FcStrListFirst (list);
  ^
../src/480391f31f-4c049f225c.clean/pango/pangofc-fontmap.c:2535:3: note: did you mean 'FcStrListNext'?
/Library/Frameworks/Mono.framework/Headers/fontconfig/fontconfig.h:959:1: note: 'FcStrListNext' declared here
FcStrListNext (FcStrList *list);
^
../src/480391f31f-4c049f225c.clean/pango/pangofc-fontmap.c:2545:10: warning: cast from 'gchar *' (aka 'char *') to 'PangoLanguage **' (aka 'struct _PangoLanguage **') increases required alignment from 1 to 8 [-Wcast-align]
  return (PangoLanguage **) g_array_free (langs, FALSE);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/480391f31f-4c049f225c.clean/pango/pangofc-fontmap.c:2659:10: error: implicit declaration of function 'FcWeightToOpenTypeDouble' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  return FcWeightToOpenTypeDouble (fc_weight);
         ^
../src/480391f31f-4c049f225c.clean/pango/pangofc-fontmap.c:2947:17: warning: cast from 'gchar *' (aka 'char *') to 'int *' increases required alignment from 1 to 4 [-Wcast-align]
              *sizes = (int *) size_array->data;
                       ^~~~~~~~~~~~~~~~~~~~~~~~
2 warnings and 3 errors generated.
```
Can you try to remove `-Wimplicit-function-declaration` and  `-Werror=implicit` from meson.build in source codes and add `-Wno-error=implicit-function-declaration`. Then rebuild this?

Note: Before update changes, please use `vcpkg install pango --editable` to rebuild first. 
```
elif cc.get_id() == 'gcc' or cc.get_id() == 'clang'
  test_c_only_flags = [
    '-Wimplicit-function-declaration', //Remove this line
    '-Wstrict-prototypes',
    '-Wmissing-prototypes',
    '-Wnested-externs',
    '-Wold-style-definition',
    '-Wno-int-conversion',
    '-Wno-discarded-qualifiers',
    '-Werror=implicit', //Remove this line
    '-Werror=pointer-to-int-cast',
    '-Wno-error=implicit-function-declaration', //Add this line
  ]
```

**Comment by acromarcopolo** [comment]
Sorry I need clarification/confirmation:
- there are many "meson.build" files. Do you mean the one in my
~/gtk/source?  at least there,i found a piece of code like you present
- are you saying the order of operation is
1/ run vcpkg install pango --editable
2/ do the changes in meson.build
3/ run ./vcpkg install gtk again
or 2/ 1/ 3/ ?

**Comment by NancyLi1013** [comment]
```
1/ run vcpkg install pango --editable
2/ do the changes in meson.build   // in pango source
```
Please make sure you can install pango successfully, then rebuild other ports.

**Comment by acromarcopolo** [comment]
step 1/ fails:
[stdout-x64-osx.log](https://github.com/microsoft/vcpkg/files/6560096/stdout-x64-osx.log)
[package-x64-osx-dbg-out.log](https://github.com/microsoft/vcpkg/files/6560097/package-x64-osx-dbg-out.log)
(log ends with `ninja: build stopped: subcommand failed.`)

**Comment by NancyLi1013** [comment]
Can you try to add `#include <fontconfig/fontconfig.h>` to `../src/480391f31f-4c049f225c.clean/pango/pangofc-fontmap.c`?

Since `FcWeightFromOpenTypeDouble`, `FcStrListFirst`, `FcWeightToOpenTypeDouble` these functions are defined in `#include <fontconfig/fontconfig.h>`, but the header is not included in the file.

**Comment by acromarcopolo** [comment]
done, and also in `../src/480391f31f-4c049f225c/pango/pangofc-fontmap.c` (without the `.clean`) after a first fail.
`./vcpkg install pango --editable`  still fails:
[stdout-x64-osx.log](https://github.com/microsoft/vcpkg/files/6575743/stdout-x64-osx.log)
[package-x64-osx-dbg-out.log](https://github.com/microsoft/vcpkg/files/6575744/package-x64-osx-dbg-out.log)

**Comment by JackBoosY** [comment]
```
../src/480391f31f-4c049f225c/pango/pangofc-fontmap.c:1780:10: error: implicit declaration of function 'FcWeightFromOpenTypeDouble' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
  return FcWeightFromOpenTypeDouble (pango_weight);
         ^
```
Try to use compiler option `c11`.

**Comment by harrynull** [comment]
I also had this issue but for what it worth it is because I didn't install freetype2.

**Comment by beached** [comment]
So adding the <fontconfig/fontconfig.h> did not work, neither did removing the warnings on implicit(I removed all of them from the ./buildtrees/pange

What fixed it was adding the following in `./buildtrees/pango/src/480391f31f-4c049f225c/pango/pangofc-fontmap.c`

```cpp
extern double FcWeightFromOpenTypeDouble(double);
extern void FcStrListFirst (FcStrList *);
extern double FcWeightToOpenTypeDouble (double);```

Afterwards it built correctly.  Kind of confusing, when I checked the fontconfig folder, the fontconfig.h file had the declarations and I added that include right below <math.h>

**Comment by beached** [comment]
Is there any other info needed to test/fix?

**Comment by JackBoosY** [comment]
`FcWeightFromOpenTypeDouble` is used in _pangofc-fontmap.c_ line 1778, not 1780.
So I think pango was updated before, can you please confirm whether this issue still exist?

**Comment by JackBoosY** [comment]
We don't have enough information to solve this issue, unfortunately; if you have any more information to help us solve this, please reopen!

**Comment by beached** [comment]
Sorry, I put this info into another related issue.  I found that the issue is that framwork header paths are taking priority over the ones specified in cmake/vcpkg.  In the case here, I was able to fix the issue by uninstalling Mono/VS Mac which included FontConfig that was not compatible.  This was the source of the missing declarations.  I couldn't find a way to remove frameworks from the search paths though
